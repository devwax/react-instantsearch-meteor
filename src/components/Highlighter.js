import PropTypes from 'prop-types';
import React from 'react';

export default function Highlighter({
  hit,
  attributeName,
  highlight,
  highlightProperty,
  tagName,
}) {
  const parsedHighlightedValue = highlight({
    hit,
    attributeName,
    highlightProperty,
  });
  const reactHighlighted = parsedHighlightedValue.map((v, i) => {
    const key = `split-${i}-${v.value}`;
    if (!v.isHighlighted) {
      return (
        <span key={key} className="ais-Highlight__nonHighlighted">
          {v.value}
        </span>
      );
    }
    const HighlightedTag = tagName ? tagName : 'em';
    return (
      <HighlightedTag key={key} className="ais-Highlight__highlighted">
        {v.value}
      </HighlightedTag>
    );
  });
  return <span className="ais-Highlight">{reactHighlighted}</span>;
}

Highlighter.propTypes = {
  hit: PropTypes.object.isRequired,
  attributeName: PropTypes.string.isRequired,
  highlight: PropTypes.func.isRequired,
  highlightProperty: PropTypes.string.isRequired,
  tagName: PropTypes.string,
};
