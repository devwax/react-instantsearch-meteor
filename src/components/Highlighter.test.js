/* eslint-env jest, jasmine */

import React from 'react';
import renderer from 'react-test-renderer';

import Highlighter from './Highlighter';
import parseAlgoliaHit from '../core/highlight';

describe('Highlighter', () => {
  it('parses an highlighted attribute of hit object', () => {
    const hitFromAPI = {
      objectID: 0,
      deep: { attribute: { value: 'awesome highlighted hit!' } },
      _highlightProperty: {
        deep: {
          attribute: {
            value: {
              value:
                'awesome <ais-highlight>hi</ais-highlight>ghlighted <ais-highlight>hi</ais-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = ({ hit, attributeName, highlightProperty }) =>
      parseAlgoliaHit({
        preTag: '<ais-highlight>',
        postTag: '</ais-highlight>',
        attributeName,
        hit,
        highlightProperty,
      });

    const tree = renderer.create(
      <Highlighter
        attributeName="deep.attribute.value"
        hit={hitFromAPI}
        highlight={highlight}
        highlightProperty="_highlightProperty"
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('renders a hit with a custom tag correctly', () => {
    const hitFromAPI = {
      objectID: 0,
      deep: { attribute: { value: 'awesome highlighted hit!' } },
      _highlightProperty: {
        deep: {
          attribute: {
            value: {
              value:
                'awesome <ais-highlight>hi</ais-highlight>ghlighted <ais-highlight>hi</ais-highlight>t!',
              fullyHighlighted: true,
              matchLevel: 'full',
              matchedWords: [''],
            },
          },
        },
      },
    };

    const highlight = ({ hit, attributeName, highlightProperty }) =>
      parseAlgoliaHit({
        preTag: '<ais-highlight>',
        postTag: '</ais-highlight>',
        attributeName,
        hit,
        highlightProperty,
      });

    const tree = renderer.create(
      <Highlighter
        attributeName="deep.attribute.value"
        hit={hitFromAPI}
        highlight={highlight}
        highlightProperty="_highlightProperty"
        tagName="mark"
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
