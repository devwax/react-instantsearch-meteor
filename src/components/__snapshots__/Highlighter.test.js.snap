// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Highlighter parses an highlighted attribute of hit object 1`] = `
<span
  className="ais-Highlight"
>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    awesome 
  </span>
  <em
    className="ais-Highlight__highlighted"
  >
    hi
  </em>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    ghlighted 
  </span>
  <em
    className="ais-Highlight__highlighted"
  >
    hi
  </em>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    t!
  </span>
</span>
`;

exports[`Highlighter renders a hit with a custom tag correctly 1`] = `
<span
  className="ais-Highlight"
>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    awesome 
  </span>
  <mark
    className="ais-Highlight__highlighted"
  >
    hi
  </mark>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    ghlighted 
  </span>
  <mark
    className="ais-Highlight__highlighted"
  >
    hi
  </mark>
  <span
    className="ais-Highlight__nonHighlighted"
  >
    t!
  </span>
</span>
`;
